### ISSUE TITLE
API compatibility with old plugin

### ISSUE THREAD

**Issue by DPUkyle** [KengoTODA/spotbugs-gradle-plugin-v2#39]
Thanks for rewriting the plugin; this is really great!

I replaced `com.github.spotgugs:spotbugs-gradle-plugin:1.6.5` with `gradle.plugin.com.github.spotbugs.snom:spotbugs-gradle-plugin-v2:0.4.4`. After fixing trivial package naming changes, I noticed a few incompatibilities with obscure APIs I use.

1. `SpotBugsTask` should continue to extend `SourceTask`; that way we get free filtering methods such as `exclude`.
2. Please also make `SpotBugsTask` implement `o.g.a.t.VerificationTask`; that way I can easily disable it via `tasks.withType(VerificationTask).ignoreFailures = true`
3. We also invoke `SBT#setIncludeFilter(File)` and `#setExcludeFilter(File)`; I did not see an equivalent API in the new plugin.
4. `c.g.s.s.SpotBugsTask#getReports()` now returns `NDOC<c.g.s.s.SpotBugsReport>`. I currently have a call to `c.g.s.SpotBugsTask#getReports().getXml()`, and `getReports()` returned an instance of `c.g.s.SporBugsReports`. For my complex reporting requirements it would be nice to have the old `SpotBugsReports` type.
5. Previously I called `c.g.s.SpotBugsTask#getSpotbugsClasspath()` to get the libraries; I think this is replaced by resolving the `spotbugs` configuration. Does that sound right?
6. `SpotBugsExtension#setToolVersion(String)` is gone; workaround is same as above, correct?
7. Please make `c.g.s.s.SpotBugsPlugin#CONFIG_NAME` public.

Finally, please confirm the default version of the spotbugs tool that is provided if the user does not specify. Maybe a table on the main README would be great.

My other advice/request is, please don't write the main classes in Groovy. Groovy is great for tests and I love the Spock framework, but I prefer the type-safety and performance of Java.

**Comment by KengoTODA** [comment]
> 1. SpotBugsTask should continue to extend SourceTask; that way we get free filtering methods such as exclude.

Do you have any good idea to realize it with Android support? Then it's awesome.

> 2. Please also make SpotBugsTask implement o.g.a.t.VerificationTask

Done.

> 3. We also invoke SBT#setIncludeFilter(File) and #setExcludeFilter(File); I did not see an equivalent API in the new plugin.

The following code works, could you have a try?

```groovy
spotbugsMain {
    includeFilter = file('...')
}
```

> 4. c.g.s.s.SpotBugsTask#getReports() now returns NDOC<c.g.s.s.SpotBugsReport>. 

Could you give me a build script snippet to know your usage?
To remove dependency on internal API, I think it's impossible to keep providing completely same interface. So I want to know your purpose clearly.

> 5. Peviously I called c.g.s.SpotBugsTask#getSpotbugsClasspath() to get the libraries; I think this is replaced by resolving the spotbugs configuration. Does that sound right?

It's not the same, I will rename existing method to keep this backward compatibility. 

> 6. SpotBugsExtension#setToolVersion(String) is gone; workaround is same as above, correct?

Yes use `spotbugs` configuration to specify the SpotBugs version to use.

> 7. Please make c.g.s.s.SpotBugsPlugin#CONFIG_NAME public.

Done.

> Finally, please confirm the default version of the spotbugs tool that is provided if the user does not specify. Maybe a table on the main README would be great.

I will handle it when the release becomes stable.

**Comment by DPUkyle** [comment]
Thank you for the progress!

re: Android and SourceTask, let me talk to my colleague on Monday.

I'll also share the example of `SpotBugsTask#getReports()`. I feel like we can have compatibility without resorting to internal APIs.

**Comment by KengoTODA** [comment]
> I feel like we can have compatibility without resorting to internal APIs.

Thanks! In my understanding, the blocker is as follows:

1. `ReportContainer` is not an internal API, but its implementation `DefaultReportContainer` is internal. All existing code quality plugin extends this (e.g. [PmdReportsImpl](https://github.com/gradle/gradle/blob/f70668403935567579f69117bd85897c4889667d/subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/internal/PmdReportsImpl.java))


BTW I changed the implementation of `SpotBugsTask#getFirstEnabledReport()` in #42.

**Comment by DPUkyle** [comment]
OK thank you, I see the point now about the old `c.g.s.i.SpotBugsReportsImpl` makes heavy use of internal APIs. I could successfully replace usages of `SpotBugsTask#getReports().getXml()` with `getReports().getByName("xml")`. It seems fine.

By the way (maybe should be a separate issue?) I am not able to build the project from master. ~It looks like tools.jar might be required for test compilation??~ JDK 9 seems to be required by errorprone: https://github.com/tbroyer/gradle-errorprone-plugin#jdk-8-support. Also see my build scan here: https://scans.gradle.com/s/exwt7a72durmg/console-log#L69

**Comment by KengoTODA** [comment]
Yes please use Java 11 or later to build. The generated .class files is compatible with Java 8.

**Comment by KengoTODA** [comment]
`toolVersion` is supported by #48, try the version `0.4.11` please

**Comment by KengoTODA** [comment]
@DPUkyle any update do we have? can we close this issue?

**Comment by DPUkyle** [comment]
I still want to extend `SourceTask`, which is fairly trivial. I can submit a PR today, Tuesday 2020-01-21 US PT. But I do not know how to test against an android project - do you have a simple example?

**Comment by DPUkyle** [comment]
OK, since you added support for `getClasses()`, I think we can drop `SourceTask` for now.

In testing, I found #67. Once that is merged I think we can close this issue. Thank you!!

Of course, once the new plugin is more widely used in my organization there may be additional issues; let's address as they are found.

**Comment by KengoTODA** [comment]
Got it, thank you!
